File: plugins/langfuse-config/scripts/reporter.py
#!/usr/bin/env python3
"""
Langfuse 上报模块。

负责将已构建好的 turn/events 映射为 Langfuse observations。
"""

import json
import sys
from pathlib import Path

SCRIPT_DIR = Path(__file__).resolve().parent
if str(SCRIPT_DIR) not in sys.path:
    sys.path.insert(0, str(SCRIPT_DIR))

from langfuse import Langfuse, propagate_attributes

from collector import (
    Turn,
    extract_text,
    get_content,
    get_model,
    get_msg_timestamp,
    sum_usage_details,
    truncate_text,
)
from langfuse_wrapper import TimedLangfuseWrapper


def emit_turn(langfuse: Langfuse, session_id: str, turn_num: int, turn: Turn, transcript_path: Path) -> None:
    """Emit a single turn to Langfuse with ordered child observations."""
    user_text_raw = extract_text(get_content(turn.user_msg))
    user_text, user_text_meta = truncate_text(user_text_raw)
    model = get_model(turn.assistant_msgs[0])
    tool_count = sum(1 for event in turn.events if event.kind == "tool_call")
    usage_details = sum_usage_details(turn.assistant_msgs)
    timed_langfuse = TimedLangfuseWrapper(langfuse)
    trace_timestamp = get_msg_timestamp(turn.user_msg) or next(
        (event.timestamp for event in turn.events if event.timestamp is not None),
        None,
    )
    trace_end_time = next(
        (event.timestamp for event in reversed(turn.events) if event.timestamp is not None),
        trace_timestamp,
    )
    tool_result_timestamps = {
        event.tool_id: event.timestamp
        for event in turn.events
        if event.kind == "tool_result" and event.tool_id
    }

    final_assistant_text_index = -1
    for index, event in enumerate(turn.events):
        if event.kind == "assistant_text" and event.text:
            final_assistant_text_index = index

    final_assistant_text = ""
    if final_assistant_text_index >= 0:
        final_event = turn.events[final_assistant_text_index]
        final_assistant_text = final_event.text

    with propagate_attributes(
        session_id=session_id,
        trace_name=f"Claude Code - Turn {turn_num}",
        tags=["claude-code"],
    ):
        with timed_langfuse.start_as_current_observation(
            name=f"Claude Code - Turn {turn_num}",
            input={"role": "user", "content": user_text},
            completion_start_time=trace_timestamp,
            start_time=trace_timestamp,
            end_time=trace_end_time,
            metadata={
                "source": "claude-code",
                "session_id": session_id,
                "turn_number": turn_num,
                "transcript_path": str(transcript_path),
                "user_text": user_text_meta,
            },
        ) as trace_span:
            for index, event in enumerate(turn.events):
                if event.kind == "assistant_text":
                    if index == final_assistant_text_index:
                        with timed_langfuse.start_as_current_observation(
                            name="Claude Response",
                            as_type="generation",
                            model=model,
                            input={"role": "user", "content": user_text},
                            output={"role": "assistant", "content": event.text},
                            completion_start_time=event.timestamp,
                            start_time=event.timestamp,
                            end_time=event.timestamp,
                            usage_details=usage_details or None,
                            metadata={
                                "assistant_text": event.text_meta,
                                "tool_count": tool_count,
                                "usage_details": usage_details or None,
                            },
                        ):
                            pass
                    elif event.text:
                        with timed_langfuse.start_as_current_observation(
                            name="Assistant Message",
                            input={"role": "assistant", "content": event.text},
                            completion_start_time=event.timestamp,
                            start_time=event.timestamp,
                            end_time=event.timestamp,
                            metadata={
                                "assistant_text": event.text_meta,
                                "phase": "pre_tool_or_intermediate",
                                "observation_kind": "assistant_text",
                            },
                        ):
                            pass
                elif event.kind == "tool_call":
                    tool_input = event.input
                    if isinstance(tool_input, str):
                        tool_input, input_meta = truncate_text(tool_input)
                    else:
                        input_meta = None
                    tool_output = None
                    tool_output_meta = None
                    if event.tool_id and event.tool_id in turn.tool_results_by_id:
                        out_raw = turn.tool_results_by_id[event.tool_id]
                        out_str = out_raw if isinstance(out_raw, str) else json.dumps(out_raw, ensure_ascii=False)
                        tool_output, tool_output_meta = truncate_text(out_str)

                    tool_end_time = tool_result_timestamps.get(event.tool_id) or event.timestamp

                    with timed_langfuse.start_as_current_observation(
                        name=f"Tool: {event.name}",
                        as_type="tool",
                        input=tool_input,
                        completion_start_time=event.timestamp,
                        start_time=event.timestamp,
                        end_time=tool_end_time,
                        metadata={
                            "tool_name": event.name,
                            "tool_id": event.tool_id,
                            "input_meta": input_meta,
                            "output_meta": tool_output_meta,
                        },
                    ) as tool_obs:
                        tool_obs.update(output=tool_output)
                elif event.kind == "tool_result":
                    with timed_langfuse.start_as_current_observation(
                        name="Tool Result",
                        output=event.output,
                        completion_start_time=event.timestamp,
                        start_time=event.timestamp,
                        end_time=event.timestamp,
                        metadata={
                            "tool_id": event.tool_id,
                            "output_meta": event.output_meta,
                            "observation_kind": "tool_result",
                        },
                    ):
                        pass

            trace_span.update(output={"role": "assistant", "content": final_assistant_text})

File: plugins/langfuse-config/scripts/reporter.js
#!/usr/bin/env node

const crypto = require("node:crypto");
const { ObservationType } = require("@langfuse/core");

const {
  extractText,
  getContent,
  getMessageTimestamp,
  getModel,
  sumUsageDetails,
  truncateText,
} = require("./collector");

function makeId(parts) {
  return crypto.createHash("sha256").update(parts.join("::"), "utf8").digest("hex").slice(0, 32);
}

function toIso(value) {
  return value instanceof Date ? value.toISOString() : undefined;
}

function buildIngestionBatch({ sessionId, turnNum, turn, transcriptPath }) {
  const userTextRaw = extractText(getContent(turn.userMsg));
  const [userText, userTextMeta] = truncateText(userTextRaw);
  const model = getModel(turn.assistantMsgs[0] || {});
  const usageDetails = sumUsageDetails(turn.assistantMsgs);
  const traceTimestamp =
    getMessageTimestamp(turn.userMsg) ||
    turn.events.find((event) => event.timestamp)?.timestamp ||
    new Date();
  const traceEndTime =
    [...turn.events].reverse().find((event) => event.timestamp)?.timestamp || traceTimestamp;
  const traceId = makeId([sessionId, String(turnNum), transcriptPath]);

  let finalAssistantText = "";
  let finalAssistantTextMeta = null;
  let finalAssistantTimestamp = traceEndTime;
  let finalAssistantIndex = -1;

  turn.events.forEach((event, index) => {
    if (event.kind === "assistant_text" && event.text) {
      finalAssistantIndex = index;
      finalAssistantText = event.text;
      finalAssistantTextMeta = event.textMeta || null;
      finalAssistantTimestamp = event.timestamp || traceEndTime;
    }
  });

  const batch = [
    {
      id: makeId([traceId, "trace-create"]),
      type: "trace-create",
      timestamp: toIso(traceTimestamp),
      body: {
        id: traceId,
        timestamp: toIso(traceTimestamp),
        name: `Claude Code - Turn ${turnNum}`,
        input: { role: "user", content: userText },
        output: { role: "assistant", content: finalAssistantText },
        sessionId,
        metadata: {
          source: "claude-code",
          session_id: sessionId,
          turn_number: turnNum,
          transcript_path: transcriptPath,
          user_text: userTextMeta,
        },
        tags: ["claude-code"],
      },
    },
  ];

  turn.events.forEach((event, index) => {
    const eventId = makeId([traceId, String(index), event.kind]);
    if (event.kind === "assistant_text") {
      if (index === finalAssistantIndex) {
        batch.push({
          id: makeId([eventId, "generation-create"]),
          type: "generation-create",
          timestamp: toIso(event.timestamp || finalAssistantTimestamp),
          body: {
            id: eventId,
            traceId,
            name: "Claude Response",
            startTime: toIso(event.timestamp || finalAssistantTimestamp),
            endTime: toIso(event.timestamp || finalAssistantTimestamp),
            completionStartTime: toIso(event.timestamp || finalAssistantTimestamp),
            model,
            input: { role: "user", content: userText },
            output: { role: "assistant", content: event.text },
            metadata: {
              assistant_text: finalAssistantTextMeta,
              tool_count: turn.events.filter((item) => item.kind === "tool_call").length,
              usage_details: Object.keys(usageDetails).length > 0 ? usageDetails : null,
            },
            usageDetails: Object.keys(usageDetails).length > 0 ? usageDetails : undefined,
          },
        });
        return;
      }

      batch.push({
        id: makeId([eventId, "event-create"]),
        type: "event-create",
        timestamp: toIso(event.timestamp || traceTimestamp),
        body: {
          id: eventId,
          traceId,
          name: "Assistant Message",
          startTime: toIso(event.timestamp || traceTimestamp),
          input: { role: "assistant", content: event.text },
          metadata: {
            assistant_text: event.textMeta || null,
            phase: "pre_tool_or_intermediate",
            observation_kind: "assistant_text",
          },
        },
      });
      return;
    }

    if (event.kind === "tool_call") {
      batch.push({
        id: makeId([eventId, "span-create"]),
        type: "span-create",
        timestamp: toIso(event.timestamp || traceTimestamp),
        body: {
          id: eventId,
          traceId,
          name: `Tool: ${event.name}`,
          startTime: toIso(event.timestamp || traceTimestamp),
          endTime: toIso(event.timestamp || traceTimestamp),
          input: event.input,
          output: event.toolId ? turn.toolResultsById[event.toolId] : undefined,
          metadata: {
            tool_name: event.name,
            tool_id: event.toolId,
          },
          level: "DEFAULT",
          type: ObservationType.Tool,
        },
      });
      return;
    }

    if (event.kind === "tool_result") {
      batch.push({
        id: makeId([eventId, "event-create"]),
        type: "event-create",
        timestamp: toIso(event.timestamp || traceTimestamp),
        body: {
          id: eventId,
          traceId,
          name: "Tool Result",
          startTime: toIso(event.timestamp || traceTimestamp),
          output: event.output,
          metadata: {
            tool_id: event.toolId,
            output_meta: event.outputMeta || null,
            observation_kind: "tool_result",
          },
        },
      });
    }
  });

  return batch;
}

async function emitTurn(langfuse, sessionId, turnNum, turn, transcriptPath) {
  const batch = buildIngestionBatch({
    sessionId,
    turnNum,
    turn,
    transcriptPath: String(transcriptPath),
  });
  await langfuse.api.ingestion.batch({ batch });
}

module.exports = {
  buildIngestionBatch,
  emitTurn,
};
